src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module
+import linear_algebra.quotient
+import ring_theory.ideal.quotient
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `torsion R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0`.
+* `torsion' R M` : the torsion submoule, containing all elements `x` of `M` such that  `a • x = 0`
+  for some non-zero `a` in `R`.
+
+## Main statements
+
+* `torsion' R M` is a submodule when `R` is a domain.
+* `torsion R M a` can be viewed as a `(R ⧸ R·a)`-module.
+* `quot_torsion_is_torsion_free` : `(M ⧸ torsion' R M)` is a torsion-free module, that is, there is
+  no non-zero `a`, `x` such that `a • x = 0`.
+
+## Notation
+
+* The notions are defined for a `comm_semiring R` and a `module R M`. Some additional hypotheses on
+  `R` and `M`are required by some lemmas.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+
+section defs
+variables (R M : Type*) [comm_semiring R] [add_comm_monoid M] [module R M] (a : R)
+
+/-- The `a`-torsion submodule, for `a` in `R` -/
+def torsion : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
+
+/-- The torsion submodule, only defined when `R` is a domain. -/
+def torsion' [no_zero_divisors R] [nontrivial R] : submodule R M :=
+{ carrier := { x | ∃ a : R, a • x = 0 ∧ a ≠ 0 },
+  zero_mem' := ⟨1, smul_zero _, one_ne_zero⟩,
+  add_mem' := λ x y ⟨a, hx, ha⟩ ⟨b, hy, hb⟩,
+    ⟨b * a,
+      by rw [smul_add, ← smul_smul, mul_comm, ← smul_smul, hx, hy, smul_zero, smul_zero, add_zero],
+      mul_ne_zero hb ha⟩,
+  smul_mem' := λ a x ⟨b, h, hb⟩, ⟨b, by rw [smul_smul, mul_comm, ← smul_smul, h, smul_zero], hb⟩ }
```

Comment:
Have you tried removing `[no_zero_divisors R]`? Here is my first attempt, which certainly is not optimal, but does the job here (two errors later in the file would have to be fixed though). Your original definition could then be added as a lemma with an equivalent formulation.

```suggestion
def torsion' [nontrivial R] : submodule R M :=
{ carrier := { x | ∃ a : non_zero_divisors R, a • x = 0 },
  zero_mem' := ⟨1, smul_zero _⟩,
  add_mem' := λ x y ⟨a, hx⟩ ⟨b, hy⟩,
    ⟨b * a,
      by rw [smul_add, ← smul_smul, mul_comm, ← smul_smul, hx, hy, smul_zero, smul_zero, add_zero]⟩,
  smul_mem' := λ a x ⟨b, h⟩, ⟨b, begin have h' : (b : R) • x = 0, exact h, have h'' : (b : R) • a • x = 0, { rw [smul_smul, mul_comm, ←smul_smul, h', smul_zero]}, exact h'' end ⟩ }
```

Reply:
Is this actually better? Your `torsion' R M` is just the old `torsion' (non_zero_divisors R) M`, so I'd perhaps argue your version is less general.

Reply:
It is at least what is usually used in algebra. If you don't require `no_zero_divisors` then the other thing is not a submodule, see e.g. [this post](https://math.stackexchange.com/q/1844567/15416).

Reply:
But `submodule R M` is defined only for a `semiring R` and `module R M`, despite it needing technically only `add_comm_monoid M` and `has_scalar R M`. The definition of `torsion' R M` further requires `monoid R` (or `monoid_with_zero R`) and `distrib_mul_action R M`.

Reply:
So, at which point we should generalise ?